### ISSUE TITLE
Reduce number of Docker image layers introduced through hotfixes

### ISSUE THREAD

**Issue by tomudding** [overleaf/overleaf#1242]
The hotfixes patch many files, but for each patch, a `COPY`/`ADD` and `RUN` is executed. This means that each patch requires at least two layers in the new (final) Docker image. Although the _hard_ limit of 127 layers seems to have been removed, some systems simply cannot handle more layers.

With the recent hotfixes in the `5.0.x` branch, the latest `5.0.5` image sits at 104 layers. This is not yet the old limit, but as more hotfixes are released we will get close to it. If you have your own tweaks to Overleaf, you can easily reach more than 127 layers. I experienced this when I tried to update our version to `5.0.5`. While this was partly due to extra overhead in the way we patch Overleaf, it was also caused by the excessive amount of unnecessary layers.

A solution to this problem would be to group `COPY`/`ADD` statements per patch into a single `COPY` and then have a corresponding `RUN`. For example, for `5.0.5` we could change the following

https://github.com/overleaf/overleaf/blob/658be2e7e11ca8b8134f6228e5aed697d490e98b/server-ce/hotfix/5.0.5/Dockerfile#L1-L12

to (comments removed for brevity)

```Dockerfile
COPY pr_18570.patch pr_18819.patch ./
RUN patch -p1 < pr_18570.patch && rm pr_18570.patch \
    && patch -p1 < pr_18819.patch && rm pr_18819.patch \
    && node genScript compile | bash
```

to go from 5 layers to just 2 layers (a 60% improvement).

For more complicated patches (which are not in the current working directory), the `RUN` can be used to move the files to the correct location _or_ -which I think is better to do- use the appropriate `-d` option for the `patch` command to specify the correct directory (this happens for a few of the existing patches). And be smart about organising the patches. If it is **not** necessary to do `node genScript compile | bash` between patches, then you can probably do that at the end, but if it is necessary, then of course you want to split the `RUN`s to improve layer caching.

Obviously, this reduces readability, so you'll have to format the comments differently to make sure it's clear what patches what, but it helps enormously to reduce the number of unnecessary layers in the Docker images for release series with lots of hotfixes.

**Comment by das7pad** [comment]
Hello,

Thank you for raising awareness on the high number of layers in our images. We are building/testing these images incrementally as we prepare hotfix releases and ultimately make use of a lot of layers as the release grows. Which of your systems does not handle more than 127 layers?
Have you looked into "squashing" our image ahead of extending it? (This is not something we officially support, similarly to how we do not support making code changes to the application in our Server Pro offering.)

Greetings,
Jakob

**Comment by tomudding** [comment]
We hit the limit somewhere in our Portainer <-> Nexus Repository connection, which broke once we reached ~155 layers. For us, it was easier to start squashing the image, as you mentioned, than having to figure out which of the two systems (or the combination of them) was the actual issue.

Some time has passed, but I believe that for `5.0.7` in combination with our patches, we reached almost 180 layers without squashing. Nowadays, we have improved our patch process considerably, which also means we introduce fewer layers. In addition, we squash all hotfixes in a release series; first, as I sent in my original post, we did that by combining more `RUN`s, but now we indeed do it with the `--squash` option.

I still think the number of layers per hotfix can be reduced, but I understand it is not always convenient or a priority. So, I think this issue is more for people who may have similar problems as we did, but with the information we have now, they can build an appropriate solution (either by manually squashing layers or by using the `--squash` build option).
